### PR TITLE
JS Components: Convert icons to TypeScript

### PR DESCRIPTION
**Pull request by manzoorwanijk**
#### Changes proposed in this Pull Request:
* This PR converts all the icons to TypeScript.
* It also converts the icons' stories to TS

#### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
* Checkout the PR
* Run `cd projects/js-packages/storybook && pnpm run storybook:dev`
* Confirm that all the stories of all the icons work as expected
* Check the consumers of icons and related components and confirm that they work as expected
* Check for any regression

Screenshot 2022-07-27 at 10 52 10 AM

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - 0-as-1202276314263999

### CODE REVIEW

File: projects/js-packages/components/components/icons/index.tsx [hunk 1]
```diff
@@ -241,28 +244,43 @@ const iconsMap = {
...
+
 /**
  * Return icon component by slug.
  *
  * @param {string} slug       - Icon slug.
- * @returns {React.Component}   Icon component.
+ * @returns {React.ComponentType<BaseIconProps>}   Icon component.
  */
-export function getIconBySlug( slug ) {
+export function getIconBySlug< Slug extends IconSlug >( slug: IconSlug ): IconsMap[ Slug ] {
```
Comment (renatoagds) [inline projects/js-packages/components/components/icons/index.tsx]:
``` suggestion
export function getIconBySlug< Slug extends IconSlug >( slug: Slug ): IconsMap[ Slug ] {
```

File: projects/js-packages/components/components/icons/index.tsx [hunk 2]
```diff
@@ -241,28 +244,43 @@ const iconsMap = {
...
 };
 
+const iconsMap = {
+	...jetpackIcons,
+	...socialIcons,
+};
+
+export type IconsMap = typeof iconsMap;
+
+export type IconSlug = keyof IconsMap;
```
Comment (renatoagds) [inline projects/js-packages/components/components/icons/index.tsx:264]: Why do we need to export them?

File: projects/js-packages/components/components/icons/index.tsx (hunk 1 above)
Comment (manzoorwanijk) [reply projects/js-packages/components/components/icons/index.tsx]: Nice catch! Thanks!

File: projects/js-packages/components/components/icons/index.tsx (hunk 2 above)
Comment (manzoorwanijk) [reply projects/js-packages/components/components/icons/index.tsx:264]: I need those in TS conversion of `ProductOffer` component for better/strict typings.

File: projects/js-packages/components/components/icons/index.tsx [hunk 3]
```diff
@@ -241,28 +244,43 @@ const iconsMap = {
...
+
 /**
  * Return icon component by slug.
  *
  * @param {string} slug       - Icon slug.
- * @returns {React.Component}   Icon component.
+ * @returns {React.ComponentType<BaseIconProps>}   Icon component.
  */
-export function getIconBySlug( slug ) {
+export function getIconBySlug< Slug extends IconSlug >( slug: Slug ): IconsMap[ Slug ] {
```
Comment (dhasilva) [inline projects/js-packages/components/components/icons/index.tsx:272]: `getIconBySlug.displayName` returns `getIconBySlug`, adding an icon to the _stories_ that just shows "getBySlug" in the screenshot.

Screenshot_2022-07-26_15-15-24

Comment (dhasilva) [reply projects/js-packages/components/components/icons/index.tsx:272]: Previously it returned `undefined` and I'm not sure why this changed. Any ideas?

Comment (manzoorwanijk) [reply projects/js-packages/components/components/icons/index.tsx:272]: Good catch @dhasilva! I have no idea why storybook does that. Maybe it parses JS and TS files differently. Anyways, I have applied a fix to not render that as an icon.

**Review by retrofox — APPROVED** [review]
LGTM

**Comment by manzoorwanijk** [comment]
r249674-wpcom

**Pull request merged**